### PR TITLE
feat(medusa): Stock location module

### PR DESCRIPTION
**Pull request by carlos-r-l-rodrigues**
What:
Module to manage stock location

Fixes CORE-916

### CODE REVIEW

File: scripts/build-openapi.js
Comment (carlos-r-l-rodrigues) [inline scripts/build-openapi.js:14]: Note:
Adding `./packages/medusa/src/types` to `swaggerInline` was the only change in this file.

File: packages/medusa/src/migrations/1666251508718-stock_location.ts
Comment (carlos-r-l-rodrigues) [inline packages/medusa/src/migrations/1666251508718-stock_location.ts:1]: todo: remove this migration once https://github.com/medusajs/medusa/pull/2883 is merged

File: packages/medusa/src/api/routes/admin/stock-locations/update-stock-location.ts
Comment (pKorsholm) [inline packages/medusa/src/api/routes/admin/stock-locations/update-stock-location.ts:162]: suggestion(non-blocking): extract as separate OAS object

File: packages/medusa/src/loaders/module-definitions/definitions.ts
Comment (pKorsholm) [inline packages/medusa/src/loaders/module-definitions/definitions.ts:7]: I think this should still be `false` as long as we keep developing before release since we have opted for not having a dedicated feature flag. @srindom wdyt?

File: packages/stock-location/jest.config.js
Comment (pKorsholm) [inline packages/stock-location/jest.config.js:11]: rm

File: packages/stock-location/src/services/stock-location.ts
Comment (pKorsholm) [inline packages/stock-location/src/services/stock-location.ts:1]: suggestion: add jsdocs

**Review by pKorsholm — COMMENTED** [review]
Very nice! I only have a couple of suggestions but nothing blocking 😄

File: packages/medusa/src/api/routes/admin/stock-locations/create-stock-location.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/api/routes/admin/stock-locations/create-stock-location.ts:127]: I thought about it, but `class StockLocationAddress` isn't exported so I was unsure what to do.
Also thought about reference it to `StockLocationAddressInput` defined on `/types`. That could be ok, considering the input will have to respect that definition.

File: packages/medusa/src/api/routes/admin/stock-locations/list-stock-locations.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/api/routes/admin/stock-locations/list-stock-locations.ts:61]: good call. That is replicated in multiple places

File: packages/medusa/src/api/routes/admin/stock-locations/update-stock-location.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/api/routes/admin/stock-locations/update-stock-location.ts:162]: referencing StockLocationAddressInput would be ok?

File: packages/medusa/src/services/sales-channel-location.ts
Comment (srindom) [inline packages/medusa/src/services/sales-channel-location.ts:1]: todo (this-file): add JSDoc

File: packages/stock-location/src/services/stock-location.ts
Comment (srindom) [reply packages/stock-location/src/services/stock-location.ts:1]: this is todo :)

**Review by srindom — COMMENTED** [review]
A few small todos :)

**Review by srindom — APPROVED** [review]
LGTM!

**Comment by srindom** [comment]
Feel free to merge - just want to remind you about the your comment in the migration file. Do you need to delete the migration before merging?

**Comment by carlos-r-l-rodrigues** [comment]
> Feel free to merge - just want to remind you about the your comment in the migration file. Do you need to delete the migration before merging?

I removed it. I fixed a few conflicts but I will ask you @srindom and @pKorsholm to take a look at the  changes on `claim.ts` service that I did.

File: packages/medusa/src/services/claim.ts
Comment (srindom) [inline packages/medusa/src/services/claim.ts:338]: todo: I would remove this check. It is generally business users who create claims and they prefer having the control of forcing a claim through despite there not being stock at this time. E.g., if they know that the items will be restocked soon. 

I can see other use cases where merchants might not want this behavior, but would decouple the check from the claim creation and instead look toward a customization option in a future operation.

Comment (srindom) [reply packages/medusa/src/services/claim.ts:338]: One thing that may make sense to add now is reserving stock for claims. (Would still decouple this from the claim creation and add it in the API handler instead.)

**Review by srindom — COMMENTED** [review]
@carlos-r-l-rodrigues I have added notes on claim.ts

File: packages/medusa/src/services/claim.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/services/claim.ts]: that is being handled in the current flow that Phil covered. I believe it is better to have another PR to change the place where the stock is being allocated.
By the way, are we keeping `reserve` or should we rename everything to `allocate`

File: packages/medusa/src/api/routes/admin/stock-locations/create-stock-location.ts
Comment (srindom) [reply packages/medusa/src/api/routes/admin/stock-locations/create-stock-location.ts:77]: This would be an excellent use case for middleware :) 

```
const locationService = req.scope.resolve("stockLocationService")

if (!locationService) {
  // throw: Stock Locations are not enabled. Please add a Stock Location module to enable this functionality.
}

next()
```

Comment (pKorsholm) [reply packages/medusa/src/api/routes/admin/stock-locations/create-stock-location.ts:77]: `moduleRegisteredMiddleware`?

File: packages/stock-location/src/services/stock-location.ts
Comment (adrien2p) [inline packages/stock-location/src/services/stock-location.ts:256]: todo: Missing context for manager

File: packages/medusa/src/services/sales-channel-location.ts
Comment (adrien2p) [inline packages/medusa/src/services/sales-channel-location.ts:72]: todo: Depending on the end comment about the missing context for the transaction, we should pass the manager as part of the context to avoid inconsistency with the actual approach. wdyt?

Comment (adrien2p) [reply packages/medusa/src/services/sales-channel-location.ts:72]: same for the update

**Review by adrien2p — CHANGES_REQUESTED** [review]
Great works 🎉  I still have a few things though

File: packages/medusa-js/src/resources/admin/stock-locations.ts
Comment (fPolic) [inline packages/medusa-js/src/resources/admin/stock-locations.ts:24]:
```suggestion
    return this.client.request("POST", path, payload, {}, customHeaders)
```

todo: comments need to be updated

**Review by fPolic — COMMENTED** [review]
Great work!
Just a small change in the client is needed but LGTM!

File: packages/medusa/src/interfaces/services/event-bus.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/interfaces/services/event-bus.ts:2]: the class StockLocationService is using it

File: packages/medusa/src/types/stock-location.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/types/stock-location.ts:68]: These types are related to the module, if any external module wants to communicate with that, this is the type that has to be used. Maybe we can find a different name in the future to identify these types, but for now I will keep it like this.

File: packages/stock-location/src/services/stock-location.ts
Comment (carlos-r-l-rodrigues) [reply packages/stock-location/src/services/stock-location.ts:59]: Count will add an extra DB query to count the records before returning them.

File: packages/stock-location/src/migrations/schema-migrations/1665749860179-setup.ts
Comment (adrien2p) [reply packages/stock-location/src/migrations/schema-migrations/1665749860179-setup.ts:63]: I was thinking of '{}'::JSONB

File: packages/stock-location/src/services/stock-location.ts
Comment (carlos-r-l-rodrigues) [reply packages/stock-location/src/services/stock-location.ts:150]: This will need to be properly handled by the Event Outbox pattern that @olivermrbl is planning.

File: packages/stock-location/src/services/stock-location.ts
Comment (carlos-r-l-rodrigues) [reply packages/stock-location/src/services/stock-location.ts:167]: Which consumer you mean? As this will be an independent module, a transaction orchestrator will be needed to revert failed transaction involving this module and the core.

File: packages/stock-location/src/services/stock-location.ts
Comment (adrien2p) [reply packages/stock-location/src/services/stock-location.ts:150]: Can we add a todo? My concern is that since this feature is not guarded, anybody can install the module and end up with discrepancies. Do we have a warning somewhere ? I might have missed it

File: packages/medusa/src/api/routes/admin/stock-locations/index.ts
Comment (adrien2p) [inline packages/medusa/src/api/routes/admin/stock-locations/index.ts:26]: suggestion(non-blocking): This can be moved directly in the top app.use

File: packages/stock-location/src/services/stock-location.ts
Comment (carlos-r-l-rodrigues) [reply packages/stock-location/src/services/stock-location.ts:167]: only on the JS client, there is a * @experimental on the JSDoc

File: packages/medusa/src/api/middlewares/check-registered-modules.ts
Comment (adrien2p) [inline packages/medusa/src/api/middlewares/check-registered-modules.ts:7]: todo: should we add a test, I think that in order for the resolve to not throw, we need to add the option `allowUnregistered` otherwise it will not return undefined but throw instead. Can I get you to try please

**Review by adrien2p — APPROVED** [review]
LGTM 🎉 Just one thing to look at before merging please

**Pull request merged**
